Lib/fontbakery/codetesting.py

```diff
@@ -210,21 +211,20 @@ def assert_results_contain(check_results,
 
     for status, msg in check_results:
         if status not in [PASS, DEBUG] and not isinstance(msg, Message):
-            raise Exception(f"Bare Python strings no longer supported in result values.\n"
+            raise Exception(
+                f"Bare Python strings are no longer supported in result values.\n"
                 f"Please use the Message class to wrap strings and to give them"
-                f" a keyword useful for identifying them (on bug reports as well as"
-                f" in the implementation of reliable code-tests).\n"
-                f"(Bare string: '{msg}')")
+                f" a keyword useful for identifying them (in bug reports as well as"
+                f" in the implementation of reliable unit tests).\n"
+                f"(Bare string: {msg!r})")
 
-        if (status == expected_status and
-            expected_msgcode == None or
-            (isinstance(msg, Message) and msg.code == expected_msgcode)):
+        if status == expected_status:
             if isinstance(msg, Message):
```

Comment:
Thanks a lot for finding this bug!
The fix is not entirely correct, though, as it removes part of the validation logic.

This was meant to mean `IF A and (B or (C and D))` but due to the lack of parenthesis around `B or (C and D)`, I think that the code was actually computing the equivalent of `IF (A and B) or (C and D)`, which is not what we want.

**Note:**
* **A** = "we got the expected status (a FAIL, WARN, PASS, etc.)"
* **B** = "we did not expect any specific keyword (that's for when we're ensure a check got a PASS or a DEBUG status, for which keywords are currently not mandatory)"
* **C** = "we are dealing with a `Message` object, which specifies both a status and an identifier keyword"
* **D** = "we must ensure the keyword (`msg.code`) we got is identical to the one we expect"

So, this must actually be fixed this way:
```python
if (status == expected_status and
    (expected_msgcode == None
     or (isinstance(msg, Message) and msg.code == expected_msgcode)):
```

Reply:
The validation logic you mention is what's making the method not work correctly.
Here's 2 examples:
1. If the [com.google.fonts/check/gpos7](https://github.com/googlefonts/fontbakery/blob/3e08267464055a9fe6922d4460e7bb47dd74235e/Lib/fontbakery/profiles/universal.py#L1606-L1623) check never yields a FAIL status, how come [this assert for FAIL](https://github.com/googlefonts/fontbakery/blob/3e08267464055a9fe6922d4460e7bb47dd74235e/tests/profiles/universal_test.py#L938-L940) passes?
2. The `com.google.fonts/check/whitespace_glyphs` check is [overridden in the Adobe Fonts profile](https://github.com/googlefonts/fontbakery/blob/3e08267464055a9fe6922d4460e7bb47dd74235e/Lib/fontbakery/profiles/adobefonts.py#L134-L140) to yield WARN instead of FAIL for 0x00A0. However, if you change [this assert](https://github.com/googlefonts/fontbakery/blob/3e08267464055a9fe6922d4460e7bb47dd74235e/tests/profiles/adobefonts_test.py#L97) from WARN to FAIL, the test still succeeds whereas it shouldn't.

Reply:
Unless you can show me an example that fails due to these changes, I really think the extra logic is unnecessary.

Reply:
The code-tests of those checks you mentioned were indeed incorrect and should have caused the build to not pass, but the incorrect statuses were being silently accepted due to the lack of a parenthesis, as I explained above. So, yes, you found a few problems and provided a partial (yet not fully correct) fix. The correct fix is to not only update those expected statuses on those check assertions, but also to update the validation code-snippet as I said above with the correct parenthesis to address that oversight of mine on the precedence of boolean operators.

This validation of keywords is an inherent part of Font Bakery's unit-testing strategy: our code tests exercise our codebase to induce the problems and then verify that the correct statuses and keywords are emitted. Removing the validation of expected log message keywords would make our unit testing weaker / less precise. It would make it error-prone, since sometimes checks have multiple different ways to FAIL (or to WARN, etc.)

Please update the PR.

Reply:
@felipesanches The snippet you posted (copy below) has imbalanced parens. Also, comparison to `None` should not be done with `==` but with `is`/`is not` instead.
I'll give your suggested changes a try once you provide an updated snippet. I want to use it verbatim so you can have no doubts.

```python
if (status == expected_status and
    (expected_msgcode == None
     or (isinstance(msg, Message) and msg.code == expected_msgcode)):
```

Reply:
oops!

yeah, remove the very first opening parens, like this:
```python
if status == expected_status and
   (expected_msgcode == None
    or (isinstance(msg, Message) and msg.code == expected_msgcode)
   ):
```

Reply:
Great!